rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,1658 @@
+the current desired state of the cluster and how to get to that state. This is achieved through a series of
+matchers which are used to pair servers with installation targets and installers. At its core, a `version-directive`
+can be thought of as a function of the form `f(server) -> optional(target,installer)`.
+
+Installation targets are arbitrary attribute mappings that must *at least* contain `version`, but may contain
+any additional information as well. Certain metadata is understood by teleport (e.g. `fips:yes|no`,
+`arch:amd64|arm64|...`), but additional metadata (e.g. `sha256sum:12345...`) is simply passed through
+to the installer.  The target to be used for a given server is the first target that
+*is not incompatible* (i.e. no attempt to find the "most compatible" target is made).  A target is
+incompatible with a server if that server's version cannot safely upgrade/downgrade to that target version,
+*or* if the target specifies a build attribute that differs from a build attribute of the current
+installation (e.g. `fips:yes` when current build is `fips:no`). We don't require that all build attributes
+are present since not all systems require knowledge of said attributes.  It is the responsibility of
+an installer to fail if it is not provided with sufficient attributes to perform the installation
+safely (e.g. the `tuf` installer would fail if the target passed to it did not contain the expected
+length and hash data).
+
+All matchers in the `version-directive` resource are lists of matchers that are checked in sequence, with
+the first matching entry being selected. If a server matches a specific sub-directive, but no installation
+targets and/or installers in that sub-directive are compatible, that server has no defined `(target,installer)`
+tuple.
+
+Beyond matching installation targets to servers, the `version-directive` also supports some basic time
+constraints to assist in scheduling, and a `stale_after` field which will be used by teleport to determine
+if the directive is old enough to start emitting warnings about it (especially useful if directives are generated
+by external plugins which might otherwise fail silently).
+
+Example `version-directive` resource:
+
+```yaml
+# version directive is a singleton resource that is either supplied by a user,
+# or periodically generated by a version controller (e.g. tuf, plugin, etc).
+# this represents the desired state of the cluster, and is used to guide a control
+# loop that matches install targets to appropriate nodes and installers.
+kind: version-directive
+version: v1
+metadata:
+  name: version-directive
+spec:
+  nonce: 2
+  status: enabled
+  version_controller: static-config
+  confid_id: <random-value>
+  stale_after: <time>
+  not_before: <time>
+  not_after: <time>
+  directives:
+    - name: Staging
+      targets:
+        - version: 2.3.4
+          fips: yes
+        - version: 2.3.4
+          fips: no
+      installers:
```

Comment:
Do we get a match on the directive even if there's no compatible installer on the node? If not, how can one choose to only receive upgrade notifications?

Reply:
Yes.  If there are no compatible installers on the node, it still matches.  Good for notifications, as you mentioned.  Also important because we want incorrect configurations to fail rather than try to continue and possible create unexpected behaviors.